### PR TITLE
Improved Stackup Class

### PR DESCRIPTION
**Pull request by maxcapodi78**

### CODE REVIEW

**Review by Samuelopez-ansys — APPROVED** [review]
LGTM

File: pyaedt/modeler/stackup_3d.py [hunk 1]
```diff
@@ -950,6 +994,16 @@ def plating_ratio(self, val):
         else:
             self._app.logger.error("Plating has to be between 0 and 1")
 
+    @property
+    def padstacks_by_layer(self):
+        """Get the padstacks definition by layers
```
Comment (MaxJPRey) [inline pyaedt/modeler/stackup_3d.py:999]:
```suggestion
        """Get the padstacks definition by layers.
```

File: pyaedt/modeler/stackup_3d.py [hunk 2]
```diff
@@ -910,6 +916,44 @@ def __init__(self, padstack, layer_name, elevation, thickness):
...
+
+        Returns
+        -------
+        float
+        """
+        return self._antipad_radius
+
+    @antipad_radius.setter
+    def antipad_radius(self, val):
+        self._antipad_radius = val
```
Comment (MaxJPRey) [inline pyaedt/modeler/stackup_3d.py:955]:
```suggestion
    def antipad_radius(self, value):
        self._antipad_radius = value
```

File: pyaedt/modeler/stackup_3d.py [hunk 3]
```diff
@@ -910,6 +916,44 @@ def __init__(self, padstack, layer_name, elevation, thickness):
...
+
+        Returns
+        -------
+        float
+        """
+        return self._pad_radius
+
+    @pad_radius.setter
+    def pad_radius(self, val):
+        self._pad_radius = val
```
Comment (MaxJPRey) [inline pyaedt/modeler/stackup_3d.py:941]:
```suggestion
    def pad_radius(self, value):
        self._pad_radius = value
```

**Review by MaxJPRey — APPROVED** [review]
LGTM.

File: pyaedt/modeler/stackup_3d.py [hunk 4]
```diff
@@ -910,6 +916,44 @@ def __init__(self, padstack, layer_name, elevation, thickness):
...
+
+        Returns
+        -------
+        str
+        """
+        return self._layer_name
+
+    @property
+    def pad_radius(self):
+        """Get/Set the pad layer on the specified layer.
```
Comment (MaxJPRey) [inline pyaedt/modeler/stackup_3d.py:931]:
```suggestion
        """Get/Set the pad radius on the specified layer.
```

File: pyaedt/modeler/stackup_3d.py [hunk 5]
```diff
@@ -910,6 +916,44 @@ def __init__(self, padstack, layer_name, elevation, thickness):
...
+        """
+        return self._pad_radius
+
+    @pad_radius.setter
+    def pad_radius(self, val):
+        self._pad_radius = val
+
+    @property
+    def antipad_radius(self):
+        """Get/Set the antipad layer on the specified layer.
```
Comment (MaxJPRey) [inline pyaedt/modeler/stackup_3d.py:945]:
```suggestion
        """Get/Set the antipad radius on the specified layer.
```

**Comment by MaxJPRey** [comment]
@maxcapodi78 I fixed the warning for documentation, it was a small details.

File: pyaedt/hfss3dlayout.py [hunk 6]
```diff
@@ -740,6 +739,40 @@ def set_export_touchstone(self, activate, export_dir=""):
         self.odesign.DesignOptions(settings, 0)
         return True
 
+    @pyaedt_function_handler()
+    def set_meshing_settings(self, mesh_method="Phi", enable_intersections_check=True, use_alternative_fallback=True):
+
+        """Export the Touchstone file automatically if the simulation is successful.
```
Comment (MaxJPRey) [inline pyaedt/hfss3dlayout.py:745]:
```suggestion
        """Define the settings of the mesh.
```

**Review by MaxJPRey — APPROVED** [review]

File: pyaedt/modeler/stackup_3d.py (hunk 4 above)
Comment (PipKat) [reply pyaedt/modeler/stackup_3d.py:933]: When training me, Alex said that properties should be a noun-based description, with no need for "Get" or "Set". If something unusual existed, an example could be included.

File: pyaedt/hfss3dlayout.py [hunk 7]
```diff
@@ -740,6 +739,40 @@ def set_export_touchstone(self, activate, export_dir=""):
...
+
+        Parameters
+        ----------
+        mesh_method : bool
+            Mesh Method. Valid options are `"Phi"`, `"PhiPlus"` and `"Classic"`.
```
Comment (PipKat) [inline pyaedt/hfss3dlayout.py:750]:
```suggestion
        mesh_method : string
            Mesh method. The default is ``Phi``. Options are `"Phi"`, `"PhiPlus"`,
            and `"Classic"`.
```

File: pyaedt/hfss3dlayout.py [hunk 8]
```diff
@@ -740,6 +739,40 @@ def set_export_touchstone(self, activate, export_dir=""):
...
+        enable_intersections_check : bool, optional
+            Enable/Disable the alternative mesh intersections checks.
```
Comment (PipKat) [inline pyaedt/hfss3dlayout.py:752]:
```suggestion
            Whether to enable the alternative mesh intersections checks. The default is ``True``.
```

File: pyaedt/modeler/Primitives3DLayout.py [hunk 9]
```diff
@@ -1017,10 +1017,40 @@ def arg_with_dim(self, Value, sUnits=None):
...
+        self,
+        component_path,
+        number_of_terminals=1,
+        placement_layer=None,
+        component_name=None,
+        pos_x=0,
+        pos_y=0,
+        create_ports=True,
+    ):
         """Place a Hfss 3d Component in Hfss3dLayout.
```
Comment (PipKat) [inline pyaedt/modeler/Primitives3DLayout.py:1029]:
```suggestion
        """Place an HFSS 3D component in HFSS 3D Layout.
```

File: pyaedt/modeler/Primitives3DLayout.py [hunk 10]
```diff
@@ -1017,10 +1017,40 @@ def arg_with_dim(self, Value, sUnits=None):
...
 
+        Parameters
+        ----------
+        component_path : str
+            Full path to a3dcomp file.
```
Comment (PipKat) [inline pyaedt/modeler/Primitives3DLayout.py:1034]:
```suggestion
            Full path to the A3DCOMP file.
```

File: pyaedt/modeler/Primitives3DLayout.py [hunk 11]
```diff
@@ -1017,10 +1017,40 @@ def arg_with_dim(self, Value, sUnits=None):
...
+        number_of_terminals : int, optional
+            Number of ports in 3d component. Default is `1`.
```
Comment (PipKat) [inline pyaedt/modeler/Primitives3DLayout.py:1036]:
```suggestion
            Number of ports in the 3D component. The default is ``1``.
```

File: pyaedt/modeler/Primitives3DLayout.py [hunk 12]
```diff
@@ -1017,10 +1017,40 @@ def arg_with_dim(self, Value, sUnits=None):
...
+        placement_layer : str, optional
+            Layer on which the component will be placed. Default is None for Top Placement.
```
Comment (PipKat) [inline pyaedt/modeler/Primitives3DLayout.py:1038]:
```suggestion
            Layer to place the component on. The default is ``None``, in which case it is placed on top.
```

File: pyaedt/modeler/Primitives3DLayout.py [hunk 13]
```diff
@@ -1017,10 +1017,40 @@ def arg_with_dim(self, Value, sUnits=None):
...
+        component_name : str, optional
+            Name of component.
```
Comment (PipKat) [inline pyaedt/modeler/Primitives3DLayout.py:1040]:
```suggestion
            Name of the component. The default is ``None``, in which case a
            default name is assigned.
```

File: pyaedt/modeler/Primitives3DLayout.py [hunk 14]
```diff
@@ -1017,10 +1017,40 @@ def arg_with_dim(self, Value, sUnits=None):
...
+        pos_x : float, optional
+            X placement.
```
Comment (PipKat) [inline pyaedt/modeler/Primitives3DLayout.py:1042]:
```suggestion
            X placement. The default is ``0``.
```

Comment (PipKat) [inline pyaedt/modeler/Primitives3DLayout.py:1042]:
```suggestion
            X placement. The default is ``0``.
```

File: pyaedt/modeler/Primitives3DLayout.py [hunk 15]
```diff
@@ -1017,10 +1017,40 @@ def arg_with_dim(self, Value, sUnits=None):
...
+        pos_y : float, optional
+            Y placement.
```
Comment (PipKat) [inline pyaedt/modeler/Primitives3DLayout.py:1044]:
```suggestion
            Y placement. The default is ``0``.
```

File: pyaedt/modeler/stackup_3d.py [hunk 16]
```diff
@@ -910,6 +918,44 @@ def __init__(self, padstack, layer_name, elevation, thickness):
         self._antipad_radius = 2
         self._units = "mm"
 
+    @property
+    def layer_name(self):
+        """Get the padstack instance layer.
```
Comment (PipKat) [inline pyaedt/modeler/stackup_3d.py:923]:
```suggestion
        """Padstack instance layer.
```

File: pyaedt/modeler/stackup_3d.py [hunk 17]
```diff
@@ -910,6 +918,44 @@ def __init__(self, padstack, layer_name, elevation, thickness):
...
+
+        Returns
+        -------
+        str
```
Comment (PipKat) [inline pyaedt/modeler/stackup_3d.py:927]:
```suggestion
        str
            Name of the padstack instance layer.
```

File: pyaedt/modeler/stackup_3d.py [hunk 18]
```diff
@@ -910,6 +918,44 @@ def __init__(self, padstack, layer_name, elevation, thickness):
...
+        """
+        return self._layer_name
+
+    @property
+    def pad_radius(self):
+        """Get/Set the pad layer on the specified layer.
+
+        Returns
+        -------
+        float
```
Comment (PipKat) [inline pyaedt/modeler/stackup_3d.py:937]:
```suggestion
        float
            Pad radius on the specified layer.
```

File: pyaedt/modeler/stackup_3d.py [hunk 19]
```diff
@@ -910,6 +918,44 @@ def __init__(self, padstack, layer_name, elevation, thickness):
...
+    def pad_radius(self, val):
+        self._pad_radius = val
+
+    @property
+    def antipad_radius(self):
+        """Get/Set the antipad layer on the specified layer.
+
+        Returns
+        -------
+        float
```
Comment (PipKat) [inline pyaedt/modeler/stackup_3d.py:951]:
```suggestion
        float
            Antipad radius on the specified layer.
```

File: pyaedt/modeler/stackup_3d.py [hunk 20]
```diff
@@ -910,6 +918,44 @@ def __init__(self, padstack, layer_name, elevation, thickness):
...
+        """
+        return self._antipad_radius
+
+    @antipad_radius.setter
+    def antipad_radius(self, val):
+        self._antipad_radius = val
+
 
 class Padstack(object):
     """Padstack Class member of Stackup3D."""
```
Comment (PipKat) [inline pyaedt/modeler/stackup_3d.py:961]:
```suggestion
    """Provides the ``Padstack`` class member of Stackup3D."""
```

Comment (PipKat) [reply pyaedt/modeler/stackup_3d.py:961]: I'm really not sure what the original class description means. I just know that we want the description to start with a verb that ends with an "s".

File: pyaedt/modeler/stackup_3d.py [hunk 21]
```diff
@@ -950,6 +996,16 @@ def plating_ratio(self, val):
         else:
             self._app.logger.error("Plating has to be between 0 and 1")
```
Comment (PipKat) [inline pyaedt/modeler/stackup_3d.py:997]:
```suggestion
            self._app.logger.error("Plating has to be between 0 and 1.")
```

Comment (PipKat) [reply pyaedt/modeler/stackup_3d.py:997]: Messages are to be complete sentences, including punctuation. I'm not sure if this change will impact a test though.

File: pyaedt/modeler/stackup_3d.py [hunk 22]
```diff
@@ -950,6 +996,16 @@ def plating_ratio(self, val):
...
 
+    @property
+    def padstacks_by_layer(self):
+        """Get the padstacks definition by layers.
```
Comment (PipKat) [inline pyaedt/modeler/stackup_3d.py:1001]:
```suggestion
        """Get the padstack definitions by layers.
```

File: pyaedt/modeler/stackup_3d.py [hunk 23]
```diff
@@ -950,6 +996,16 @@ def plating_ratio(self, val):
...
+
+        Returns
+        -------
+        dict
```
Comment (PipKat) [inline pyaedt/modeler/stackup_3d.py:1005]:
```suggestion
        dict
            Dictionary of padstack definitions by layers.
```

File: pyaedt/modeler/stackup_3d.py [hunk 24]
```diff
@@ -950,6 +996,16 @@ def plating_ratio(self, val):
...
+        """
+        return self._padstacks_by_layer
+
     @property
     def num_sides(self):
         """Number of sides on the circle, which is 0 for a true circle.
```
Comment (PipKat) [inline pyaedt/modeler/stackup_3d.py:1011]:
```suggestion
        """Number of sides on the circle, which is ``0`` for a true circle.
```

File: pyaedt/modeler/stackup_3d.py [hunk 25]
```diff
@@ -1184,12 +1240,25 @@ def __init__(self, application, frequency=None):
...
+
         if frequency:
             self._frequency = NamedVariable(self._app, "frequency", str(frequency) + "Hz")
         else:
             self._frequency = frequency
         self._padstacks = []
 
+    @property
+    def thickness(self):
+        """Get the total stackup thickness.
```
Comment (PipKat) [inline pyaedt/modeler/stackup_3d.py:1254]:
```suggestion
        """Total stackup thickness.
```

File: pyaedt/application/Analysis3D.py [hunk 26]
```diff
@@ -895,7 +895,6 @@ def add_stackup_3d(self):
         Returns
         -------
         :class:`pyaedt.modeler.stackup_3d.Stackup3D`
-            ``True`` when delete operation is successful.
+            Stackup class.
```
[the remark this answers is not in the document]
Comment (maxcapodi78) [reply pyaedt/application/Analysis3D.py:898]: it's only a description

File: pyaedt/modeler/stackup_3d.py (hunk 4 above)
Comment (maxcapodi78) [reply pyaedt/modeler/stackup_3d.py:933]: good, I'll fix it

File: pyaedt/application/Analysis.py [hunk 27]
```diff
@@ -1569,7 +1570,8 @@ def analyze_setup(self, name, num_cores=None, num_tasks=None, num_gpu=None, acf_
                 return False
         if set_custom_dso:
             self.set_registry_key(r"Desktop/ActiveDSOConfigurations/" + self.design_type, active_config)
-        self.logger.info("Design setup %s solved correctly", name)
+        end = (time.time() - start) / 60
+        self.logger.info("Design setup %s solved correctly in %s min", name, end)
```
Comment (Alberto-DM) [inline pyaedt/application/Analysis.py:1574]:
```suggestion
        m, s = divmod(time.time() - start, 60)
        h, m = divmod(m, 60)
        self.logger.info("Design setup {} solved correctly in {:d}h {:02d}m {:02d}s".format(name, h, m, s))
```
This looks a bit better for simulations that last few seconds and for longer simulations.
It works in Ironpython and it does not require additional import.

**Review by Alberto-DM — APPROVED** [review]
LGTM. Big refactoring!

**Pull request merged**